query/functions.go

```diff
@@ -566,6 +566,76 @@ func (r *UnsignedMovingAverageReducer) Emit() []FloatPoint {
 	}
 }
 
+// ExponentialMovingAverageReducer calculates the
+// exponential moving average of the aggregated points.
+// It can aggregate both float and integer points.
+type ExponentialMovingAverageReducer struct {
+	decay float64 // the decay factor ( == 1/(age + 1) )
+	value float64 // current value of the moving average
+	count uint32  // number of aggregated samples
+	time  int64   // time of last sample
+}
+
+// NewExponentialMovingAverageReducer creates a new ExponentialMovingAverageReducer.
+func NewExponentialMovingAverageReducer(age int) *ExponentialMovingAverageReducer {
+	return &ExponentialMovingAverageReducer{
+		decay: 2 / (float64(age) + 1),
+	}
+}
+
+// exponentialMovingAverageWarmupSamples is the number of samples used to warm up
+// the exponential moving average calculation.
+//
+// For best results, the moving average should not be initialized to the
+// samples it sees immediately.  The book "Production and Operations
+// Analysis" by Steven Nahmias suggests initializing the moving average to
+// the mean of the first 10 samples.  Until the exponential moving average
+// reducer has seen this many samples, it is not "ready" to be queried for the
+// value of the moving average.
+const exponentialMovingAverageWarmupSamples = 10
+
+// aggregates aggregates a value to the reducer.
+func (r *ExponentialMovingAverageReducer) aggregate(value float64, time int64) {
+	r.count++
+	r.time = time
+	switch {
+	case r.count < exponentialMovingAverageWarmupSamples:
+		r.value += value
+	case r.count == exponentialMovingAverageWarmupSamples:
+		r.value = r.value / float64(exponentialMovingAverageWarmupSamples)
```

Comment:
This is another highly contentious point. How you warm the moving average can have a significant impact on the values at the start. Some people like using a simple moving average (what you've done), and others like jumping straight into the exponential moving average (which is my preference).

It's also possible to start emitting values without waiting for warmup to complete. The values might not be very stable, but depending on your usage, unstable data might be better than no data. Effectively you just start the algorithm with your weight as if the warmup count were 1, and then as the algorithm gets more data, you scale the weight until you reach the real warmup count.

Reply:
Again, of what importance is the accuracy of the first few values of the graph? In a dataset with highly varying data, some outliers could make the EMA graph potentially very ugly and stretched across the axis (making the rest of the graph scale weird). I think this warmup period only serves to avoid that.